### PR TITLE
[WIP] Move reserve related data to storage

### PR DESCRIPTION
**Pull request by manhlx3006**
Note: just compare to review and discuss to finalize the idea

### CODE REVIEW

File: contracts/sol6/KyberNetwork.sol [hunk 1]
```diff
@@ -939,18 +865,15 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils5, IKyberNetwork, Reentra
 
         reservesData.srcAmounts = new uint256[](numReserves);
         reservesData.rates = new uint256[](numReserves);
-        reservesData.addresses = new IKyberReserve[](numReserves);
         feesAccountedDestBps = new uint256[](numReserves);
+        reservesData.addresses = kyberStorage.getListReservesByIds(reservesData.ids);
```
Comment (ilanDoron) [inline contracts/sol6/KyberNetwork.sol:869]: better do this in line 820
update that function name from getFeeAccountedAndEntitledRebateData
to 
getReserveData

And query all data in 1 call. save a contract call.

File: contracts/sol6/KyberNetwork.sol [hunk 2]
```diff
@@ -1019,7 +942,7 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils5, IKyberNetwork, Reentra
 
         for (uint256 i = 0; i < reservesData.isEntitledRebateFlags.length; i++) {
             if (reservesData.isEntitledRebateFlags[i]) {
-                rebateWallets[_index] = reserveRebateWallet[address(reservesData.addresses[i])];
+                rebateWallets[_index] = kyberStorage.getRebateWallet(address(reservesData.addresses[i]));
```
Comment (ilanDoron) [inline contracts/sol6/KyberNetwork.sol:945]: lets do this in one contract call.

prepare a list of reserves
and query wallet list after reserve list is ready.
so we can save a few extra contract calls.

File: contracts/sol6/KyberNetwork.sol (hunk 1 above)
Comment (manhlx3006) [reply contracts/sol6/KyberNetwork.sol]: Updated, change name to getReserveData

File: contracts/sol6/KyberNetwork.sol (hunk 2 above)
Comment (manhlx3006) [reply contracts/sol6/KyberNetwork.sol]: Change to get list of rebateReserveIds, then call storage once to get list of rebate wallets

File: contracts/sol6/KyberNetwork.sol [hunk 3]
```diff
@@ -986,10 +933,11 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils5, IKyberNetwork, Reentra
         }
 
         uint256 index;
+        bytes32[] memory rebateReserveIds = new bytes32[](tradeData.numEntitledRebateReserves);
 
         // token to ether
         index = populateRebateWalletList(
```
Comment (ilanDoron) [inline contracts/sol6/KyberNetwork.sol:939]: need to update name.
maybe createRebateEntitledList

File: contracts/sol6/KyberNetwork.sol [hunk 4]
```diff
@@ -119,17 +114,10 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils5, IKyberNetwork, Reentra
...
-        bool add
-    );
-    event ListReservePairs(
-        address indexed reserve,
-        IERC20 indexed src,
-        IERC20 indexed dest,
+
+    event ListedReservesForToken(
+        IERC20 indexed token,
+        address[] indexed reserves,
```
Comment (Anyhowclick) [inline contracts/sol6/KyberNetwork.sol:120]: Remove indexed here.... Cos it'll be the hash of the array, which won't be useful for indexing:
https://solidity.readthedocs.io/en/v0.6.6/contracts.html#events

Comment (manhlx3006) [reply contracts/sol6/KyberNetwork.sol:120]: oh ok

File: contracts/sol6/IKyberStorage.sol [hunk 5]
```diff
@@ -97,10 +90,13 @@ interface IKyberStorage {
         view
         returns (bool[] memory entitledRebateArr);
 
-    function getFeeAccountedAndEntitledRebateData(bytes32[] calldata reserveIds)
+    function getReserveData(bytes32[] calldata reserveIds)
```
Comment (Anyhowclick) [inline contracts/sol6/IKyberStorage.sol:93]: Maybe getReservesData? Plural

File: contracts/sol6/KyberStorage.sol [hunk 6]
```diff
@@ -24,6 +25,7 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
     IKyberNetworkProxy[] internal kyberProxyArray;
 
     mapping(bytes32 => address[]) public reserveIdToAddresses;
+    mapping(bytes32 => address) public reserveRebateWallet;
     mapping(address => bytes32) internal reserveAddressToId;
```
Comment (Anyhowclick) [inline contracts/sol6/KyberStorage.sol:29]: Hmmm, should we have this public as well? Or keep reserveIdToAddresses internal? For consistency. Seems weird to me that one side is public, the other side internal.

File: contracts/sol6/KyberStorage.sol [hunk 7]
```diff
@@ -14,7 +15,7 @@ import "./utils/PermissionGroupsNoModifiers.sol";
  *   - Stores feeAccounted data for reserve types
  *   - Record contract changes for network, matchingEngine, feeHandler, reserves, network proxies and kyberDAO
  */
-contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
+contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers, Utils5 {
```
Comment (Anyhowclick) [inline contracts/sol6/KyberStorage.sol:18]: Utils5 is used? o.O

**Comment by Anyhowclick** [comment]
For KyberStorage, can we also have a listReservesForToken function (is approveTokenAllowanceForReserves a better name?) that directly calls the listReservesForToken on KN.sol?

File: contracts/sol6/KyberStorage.sol (hunk 6 above)
Comment (ilanDoron) [reply contracts/sol6/KyberStorage.sol:29]: so if its internal
We query using getReserveData?

File: contracts/sol6/KyberStorage.sol (hunk 7 above)
Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol:18]: Will remove it, seems we only use ETH address

File: test/sol6/kyberNetwork.js [hunk 8]
```diff
@@ -577,33 +577,48 @@ contract('KyberNetwork', function(accounts) {
...
+            })
+            Helper.assertEqual(txResult.logs[0].args.reserves.length, 2);
+            index = 0;
             for (const [key, value] of Object.entries(reserveInstances)) {
                 reserve = value.instance;
                 Helper.assertEqual(
                     zeroBN,
                     await token.allowance(tempNetwork.address, reserve.address)
                 )
+                Helper.assertEqual(txResult.logs[0].args.reserves[index], reserve.address);
```
Comment (ilanDoron) [inline test/sol6/kyberNetwork.js:620]: can use log name?
if later index changes?

File: contracts/sol6/KyberStorage.sol [hunk 9]
```diff
@@ -15,7 +14,10 @@ import "./utils/Utils5.sol";
  *   - Stores feeAccounted data for reserve types
  *   - Record contract changes for network, matchingEngine, feeHandler, reserves, network proxies and kyberDAO
  */
-contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers, Utils5 {
+contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
+    IERC20 internal constant ETH_TOKEN_ADDRESS = IERC20(
+        0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE
```
Comment (ilanDoron) [inline contracts/sol6/KyberStorage.sol:19]: didn't follow
Why can't we use utils?

Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol:19]: Should we use it just to use ETH_TOKEN_ADDRESS?

File: contracts/sol6/KyberStorage.sol [hunk 10]
```diff
@@ -375,16 +375,23 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
...
         view
         override
         returns (address[] memory reserveAddresses)
     {
         bytes32[] memory reserveIds = reservesPerTokenSrc[token];
-        reserveAddresses = new address[](reserveIds.length);
-        for(uint i = 0; i < reserveAddresses.length; i++) {
-            reserveAddresses[i] = reserveIdToAddresses[reserveIds[i]][0];
+        if (reserveIds.length == 0) {
+            return reserveAddresses ;
```
Comment (ilanDoron) [inline contracts/sol6/KyberStorage.sol:386]: can we return without init?
seems its working ok with the tests...
right?

File: test/sol6/kyberStorage.js [hunk 11]
```diff
@@ -501,92 +501,127 @@ contract('KyberStorage', function(accounts) {
...
-            Helper.assertEqualArray(result, reserveIds, "T2E should be listed");
+            Helper.assertEqualArray(result, reserveIds, "E2T should be listed");
             // delist for both side
             for (let reserve of Object.values(reserveInstances)) {
                 await kyberStorage.listPairForReserve(reserve.address, token.address, true, true, false, {from: operator});
             }
         });
 
-        it("should list T2E side only", async() => {
+        it("should list E2T side only", async() => {
```
Comment (ilanDoron) [inline test/sol6/kyberStorage.js:558]: why these changes?

File: test/sol6/kyberStorage.js [hunk 12]
```diff
@@ -501,92 +501,127 @@ contract('KyberStorage', function(accounts) {
...
+            result = await kyberStorage.getReserveAddressesPerTokenSrc(token.address, 0, 1);
+            Helper.assertEqual(result.length, zeroBN, "T2E should not be listed");
 
             result = await kyberStorage.getReservesPerTokenDest(token.address);
-            Helper.assertEqual(result[0], reserve.reserveId, "T2E should be listed");
+            Helper.assertEqual(result[0], reserve.reserveId, "E2T should be listed");
         });
 
-        it("should list E2T side only", async() => {
+        it("should list T2E side only", async() => {
```
Comment (ilanDoron) [inline test/sol6/kyberStorage.js:569]: why these changes?

File: contracts/sol6/KyberStorage.sol [hunk 13]
```diff
@@ -375,16 +375,23 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
         reserveIds = reservesPerTokenSrc[token];
     }
 
-    function getReserveAddressesPerTokenSrc(address token)
```
Comment (ilanDoron) [inline contracts/sol6/KyberStorage.sol]: can u add some @dev comment so its clear why we have start and end index.

File: contracts/sol6/KyberNetwork.sol (hunk 3 above)
Comment (manhlx3006) [reply contracts/sol6/KyberNetwork.sol]: Updated name

File: contracts/sol6/KyberStorage.sol (hunk 6 above)
Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol:31]: Updated

File: test/sol6/kyberNetwork.js (hunk 8 above)
Comment (manhlx3006) [reply test/sol6/kyberNetwork.js]: Updated

File: test/sol6/kyberStorage.js (hunk 11 above)
Comment (manhlx3006) [reply test/sol6/kyberStorage.js:558]: The listing params: `(reserve, token, eth2Token, token2Eth, add)`
so here we use: `(reserve, token, true, false, true)` means E2T is listed only right?

File: test/sol6/kyberStorage.js (hunk 12 above)
Comment (manhlx3006) [reply test/sol6/kyberStorage.js:569]: The listing params: `(reserve, token, eth2Token, token2Eth, add)`
so here we use: `(reserve, token, false, true, true)` means T2E is listed only right?

File: contracts/sol6/KyberStorage.sol (hunk 13 above)
Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol]: Added dev comment, please help to check if it is fit

File: contracts/sol6/KyberStorage.sol (hunk 10 above)
Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol:390]: Yeah it is working fine. Both in tests and also in deployed contracts

File: test/sol6/kyberNetwork.js [hunk 14]
```diff
@@ -551,90 +551,176 @@ contract('KyberNetwork', function(accounts) {
...
+            it("test list/unlist 1 reserve", async() => {
+                // list 1 reserve
+                let txResult = await tempNetwork.listReservesForToken(
+                    token.address, 0, 0, true, {from: operator}
+                )
+                expectEvent(txResult, 'ListedReservesForToken', {
+                    token: token.address,
+                    add: true
+                })
+                var eventLogs;
```
Comment (ilanDoron) [inline test/sol6/kyberNetwork.js:637]: how var all of the sudden?

File: test/sol6/kyberStorage.js (hunk 11 above)
Comment (ilanDoron) [reply test/sol6/kyberStorage.js:558]: yes
Got it.

File: test/sol6/kyberStorage.js (hunk 12 above)
Comment (ilanDoron) [reply test/sol6/kyberStorage.js:569]: OK

File: contracts/sol6/KyberMatchingEngine.sol [hunk 15]
```diff
@@ -203,11 +203,13 @@ contract KyberMatchingEngine is KyberHintHandler, IKyberMatchingEngine, Withdraw
     }
 
     function convertReserveIdToAddress(bytes32 reserveId) internal view override returns (address) {
-        return kyberStorage.convertReserveIdToAddress(reserveId);
+        address reserveAddress;
```
Comment (ilanDoron) [inline contracts/sol6/KyberMatchingEngine.sol:206]: can save additional parameter by using return value to define this parameter

File: contracts/sol6/KyberStorage.sol [hunk 16]
```diff
@@ -46,6 +72,16 @@ contract KyberStorage is IKyberStorage, PermissionGroupsNoModifiers {
         kyberNetwork = _kyberNetwork;
     }
 
+    function setRebateWallet(address reserve, address rebateWallet) external {
```
Comment (ilanDoron) [inline contracts/sol6/KyberStorage.sol:75]: seems we can set by ID

File: contracts/sol6/KyberStorage.sol (hunk 10 above)
Comment (ilanDoron) [reply contracts/sol6/KyberStorage.sol:390]: Goood

File: contracts/sol6/KyberStorage.sol (hunk 7 above)
Comment (manhlx3006) [reply contracts/sol6/KyberStorage.sol:18]: Back to use Utils5 as its size is quite small
